File: go/types/struct_test.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package types

import (
	"testing"

	"github.com/attic-labs/testify/assert"
)

func getChunks(v Value) (chunks []Ref) {
	v.WalkRefs(func(r Ref) {
		chunks = append(chunks, r)
	})
	return
}

func TestGenericStructEquals(t *testing.T) {
	assert := assert.New(t)

	s1 := NewStruct("S1", StructData{"s": String("hi"), "x": Bool(true)})
	s2 := NewStruct("S1", StructData{"s": String("hi"), "x": Bool(true)})

	assert.True(s1.Equals(s2))
	assert.True(s2.Equals(s1))
}

func TestGenericStructChunks(t *testing.T) {
	assert := assert.New(t)

	b := Bool(true)
	s1 := NewStruct("S1", StructData{"r": NewRef(b)})

	assert.Len(getChunks(s1), 1)
	assert.Equal(Bool(true).Hash(), getChunks(s1)[0].TargetHash())
}

func TestGenericStructNew(t *testing.T) {
	assert := assert.New(t)

	s := NewStruct("S2", StructData{"b": Bool(true), "o": String("hi")})
	assert.True(s.Get("b").Equals(Bool(true)))
	_, ok := s.MaybeGet("missing")
	assert.False(ok)

	s2 := NewStruct("S2", StructData{"b": Bool(false), "o": String("hi")})
	assert.True(s2.Get("b").Equals(Bool(false)))
	o, ok := s2.MaybeGet("o")
	assert.True(ok)
	assert.True(String("hi").Equals(o))
}

func TestGenericStructSet(t *testing.T) {
	assert := assert.New(t)

	s := NewStruct("S3", StructData{"b": Bool(true), "o": String("hi")})
	s2 := s.Set("b", Bool(false))

	s3 := s2.Set("b", Bool(true))
	assert.True(s.Equals(s3))

	// Changes the type
	s4 := s.Set("b", Number(42))
	assert.True(MakeStructType("S3",
		StructField{"b", NumberType, false},
		StructField{"o", StringType, false},
	).Equals(TypeOf(s4)))

	// Adds a new field
	s5 := s.Set("x", Number(42))
	assert.True(MakeStructType("S3",
		StructField{"b", BoolType, false},
		StructField{"o", StringType, false},
		StructField{"x", NumberType, false},
	).Equals(TypeOf(s5)))

	// Subtype is not equal.
	s6 := NewStruct("", StructData{"l": NewList(Number(0), Number(1), Bool(false), Bool(true))})
	s7 := s6.Set("l", NewList(Number(2), Number(3)))
	t7 := MakeStructTypeFromFields("", FieldMap{
		"l": MakeListType(NumberType),
	})
	assert.True(t7.Equals(TypeOf(s7)))
}

func TestGenericStructDelete(t *testing.T) {
	assert := assert.New(t)

	s1 := NewStruct("S", StructData{"b": Bool(true), "o": String("hi")})

	s2 := s1.Delete("notThere")
	assert.True(s1.Equals(s2))

	s3 := s1.Delete("o")
	s4 := NewStruct("S", StructData{"b": Bool(true)})
	assert.True(s3.Equals(s4))

	s5 := s3.Delete("b")
	s6 := NewStruct("S", StructData{})
	assert.True(s5.Equals(s6))
}

func assertValueChangeEqual(assert *assert.Assertions, c1, c2 ValueChanged) {
	assert.Equal(c1.ChangeType, c2.ChangeType)
	assert.Equal(EncodedValue(c1.Key), EncodedValue(c2.Key))
	if c1.NewValue == nil {
		assert.Nil(c2.NewValue)
	} else {
		assert.Equal(EncodedValue(c1.NewValue), EncodedValue(c2.NewValue))
	}
	if c1.OldValue == nil {
		assert.Nil(c2.OldValue)
	} else {
		assert.Equal(EncodedValue(c1.OldValue), EncodedValue(c2.OldValue))
	}
}

func TestStructDiff(t *testing.T) {
	assert := assert.New(t)

	assertDiff := func(expect []ValueChanged, s1, s2 Struct) {
		changes := make(chan ValueChanged)
		go func() {
			s1.Diff(s2, changes, nil)
			close(changes)
		}()
		i := 0
		for change := range changes {
			assertValueChangeEqual(assert, expect[i], change)
			i++
		}
		assert.Equal(len(expect), i, "Wrong number of changes")
	}

	vc := func(ct DiffChangeType, fieldName string, oldV, newV Value) ValueChanged {
		return ValueChanged{ct, String(fieldName), oldV, newV}
	}

	s1 := NewStruct("", StructData{"a": Bool(true), "b": String("hi"), "c": Number(4)})

	assertDiff([]ValueChanged{},
		s1, NewStruct("", StructData{"a": Bool(true), "b": String("hi"), "c": Number(4)}))

	assertDiff([]ValueChanged{vc(DiffChangeModified, "a", Bool(false), Bool(true)), vc(DiffChangeModified, "b", String("bye"), String("hi"))},
		s1, NewStruct("", StructData{"a": Bool(false), "b": String("bye"), "c": Number(4)}))

	assertDiff([]ValueChanged{vc(DiffChangeModified, "b", String("bye"), String("hi")), vc(DiffChangeModified, "c", Number(5), Number(4))},
		s1, NewStruct("", StructData{"a": Bool(true), "b": String("bye"), "c": Number(5)}))

	assertDiff([]ValueChanged{vc(DiffChangeModified, "a", Bool(false), Bool(true)), vc(DiffChangeModified, "c", Number(10), Number(4))},
		s1, NewStruct("", StructData{"a": Bool(false), "b": String("hi"), "c": Number(10)}))

	assertDiff([]ValueChanged{vc(DiffChangeAdded, "a", nil, Bool(true))},
		s1, NewStruct("NewType", StructData{"b": String("hi"), "c": Number(4)}))

	assertDiff([]ValueChanged{vc(DiffChangeAdded, "b", nil, String("hi"))},
		s1, NewStruct("NewType", StructData{"a": Bool(true), "c": Number(4)}))

	assertDiff([]ValueChanged{vc(DiffChangeRemoved, "Z", Number(17), nil)},
		s1, NewStruct("NewType", StructData{"Z": Number(17), "a": Bool(true), "b": String("hi"), "c": Number(4)}))

	assertDiff([]ValueChanged{vc(DiffChangeAdded, "b", nil, String("hi")), vc(DiffChangeRemoved, "d", Number(5), nil)},
		s1, NewStruct("NewType", StructData{"a": Bool(true), "c": Number(4), "d": Number(5)}))

	s2 := NewStruct("", StructData{
		"a": NewList(Number(0), Number(1)),
		"b": NewMap(String("foo"), Bool(false), String("bar"), Bool(true)),
		"c": NewSet(Number(0), Number(1), String("foo")),
	})

	assertDiff([]ValueChanged{},
		s2, NewStruct("", StructData{
			"a": NewList(Number(0), Number(1)),
			"b": NewMap(String("foo"), Bool(false), String("bar"), Bool(true)),
			"c": NewSet(Number(0), Number(1), String("foo")),
		}))

	assertDiff([]ValueChanged{
		vc(DiffChangeModified, "a", NewList(Number(1), Number(1)), NewList(Number(0), Number(1))),
		vc(DiffChangeModified, "b", NewMap(String("foo"), Bool(true), String("bar"), Bool(true)), NewMap(String("foo"), Bool(false), String("bar"), Bool(true))),
	},
		s2, NewStruct("", StructData{
			"a": NewList(Number(1), Number(1)),
			"b": NewMap(String("foo"), Bool(true), String("bar"), Bool(true)),
			"c": NewSet(Number(0), Number(1), String("foo")),
		}))

	assertDiff([]ValueChanged{
		vc(DiffChangeModified, "a", NewList(Number(0)), NewList(Number(0), Number(1))),
		vc(DiffChangeModified, "c", NewSet(Number(0), Number(2), String("foo")), NewSet(Number(0), Number(1), String("foo"))),
	},
		s2, NewStruct("", StructData{
			"a": NewList(Number(0)),
			"b": NewMap(String("foo"), Bool(false), String("bar"), Bool(true)),
			"c": NewSet(Number(0), Number(2), String("foo")),
		}))

	assertDiff([]ValueChanged{
		vc(DiffChangeModified, "b", NewMap(String("boo"), Bool(false), String("bar"), Bool(true)), NewMap(String("foo"), Bool(false), String("bar"), Bool(true))),
		vc(DiffChangeModified, "c", NewSet(Number(0), Number(1), String("bar")), NewSet(Number(0), Number(1), String("foo"))),
	},
		s2, NewStruct("", StructData{
			"a": NewList(Number(0), Number(1)),
			"b": NewMap(String("boo"), Bool(false), String("bar"), Bool(true)),
			"c": NewSet(Number(0), Number(1), String("bar")),
		}))
}

func TestEscStructField(t *testing.T) {
	assert := assert.New(t)
	cases := []string{
		"a", "a",
		"AaZz19_", "AaZz19_",
		"Q", "Q51",
		"AQ1", "AQ511",
		"INSPECTIONQ20STATUS", "INSPECTIONQ5120STATUS",
		"$", "Q24",
		"_content", "Q5Fcontent",
		"Few ¢ents Short", "FewQ20QC2A2entsQ20Short",
		"💩", "QF09F92A9",
		"https://picasaweb.google.com/data", "httpsQ3AQ2FQ2FpicasawebQ2EgoogleQ2EcomQ2Fdata",
	}

	for i := 0; i < len(cases); i += 2 {
		orig, expected := cases[i], cases[i+1]
		assert.Equal(expected, EscapeStructField(orig))
	}
}

func TestMakeStructTemplate(t *testing.T) {
	assert := assert.New(t)

	assertInvalidStructName := func(n string) {
		assert.Panics(func() {
			MakeStructTemplate(n, []string{})
		})
	}

	assertInvalidStructName(" ")
	assertInvalidStructName(" a")
	assertInvalidStructName("a ")
	assertInvalidStructName("0")
	assertInvalidStructName("_")
	assertInvalidStructName("0a")
	assertInvalidStructName("_a")
	assertInvalidStructName("💩")

	assertValidStructName := func(n string) {
		template := MakeStructTemplate(n, []string{})
		str := template.NewStruct(nil)
		assert.Equal(n, str.Name())
	}

	assertValidStructName("")
	assertValidStructName("a")
	assertValidStructName("A")
	assertValidStructName("a0")
	assertValidStructName("a_")
	assertValidStructName("a0_")

	assertInvalidFieldName := func(n string) {
		assert.Panics(func() {
			MakeStructTemplate("", []string{n})
		})
	}

	assertInvalidFieldName("")
	assertInvalidFieldName(" ")
	assertInvalidFieldName(" a")
	assertInvalidFieldName("a ")
	assertInvalidFieldName("0")
	assertInvalidFieldName("_")
	assertInvalidFieldName("0a")
	assertInvalidFieldName("_a")
	assertInvalidFieldName("💩")

	assertValidFieldName := func(n string) {
		MakeStructTemplate("", []string{n})
	}

	assertValidFieldName("a")
	assertValidFieldName("A")
	assertValidFieldName("a0")
	assertValidFieldName("a_")
	assertValidFieldName("a0_")

	assertInvalidFieldOrder := func(n []string) {
		assert.Panics(func() {
			MakeStructTemplate("", n)
		})
	}

	assertInvalidFieldOrder([]string{"a", "a"})
	assertInvalidFieldOrder([]string{"b", "a"})
	assertInvalidFieldOrder([]string{"a", "c", "b"})

	assertValidFieldOrder := func(n []string) {
		MakeStructTemplate("", n)
	}

	assertValidFieldOrder([]string{"a", "b"})
	assertValidFieldOrder([]string{"a", "b", "c"})

	template := MakeStructTemplate("A", []string{"a", "b"})
	str := template.NewStruct([]Value{Number(42), Bool(true)})
	assert.True(NewStruct("A", StructData{
		"a": Number(42),
		"b": Bool(true),
	}).Equals(str))
}

func TestStructWithNil(t *testing.T) {
	assert.Panics(t, func() {
		NewStruct("A", StructData{
			"a": nil,
		})
	})
	assert.Panics(t, func() {
		NewStruct("A", StructData{
			"a": Number(42),
			"b": nil,
		})
	})
}

File: js/noms/src/struct-test.js
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

// @flow

import {TestDatabase} from './test-util.js';
import Struct, {
  newStruct,
  newStructWithType,
  StructMirror,
  structDiff,
  createStructClass,
  escapeStructField,
} from './struct.js';
import {assert} from 'chai';
import {
  boolType,
  makeCycleType,
  makeListType,
  makeStructType,
  numberType,
  stringType,
} from './type.js';
import {suite, test} from 'mocha';
import {equals} from './compare.js';
import List from './list.js';
import Map from './map.js';
import Set from './set.js';
import Hash from './hash.js';

suite('Struct', () => {
  test('equals', () => {
    const data1 = {x: true, o: 'hi'};
    const s1 = newStruct('S1', data1);
    const s2 = newStruct('S1', data1);

    assert.isTrue(equals(s1, s2));
  });

  test('chunks', () => {
    const db = new TestDatabase();

    const b = true;
    const r = db.writeValue(b);
    const s1 = newStruct('S1', {r: r});
    assert.strictEqual(1, s1.chunks.length);
    assert.isTrue(equals(r, s1.chunks[0]));
    return db.close();
  });

  test('new', () => {
    const s1 = newStruct('S2', {b: true, o: 'hi'});
    assert.strictEqual(s1.b, true);
    assert.strictEqual(s1.o, 'hi');

    const s2 = newStruct('S2', {b: false, o: 'hi'});
    assert.strictEqual(s2.b, false);
    assert.strictEqual(s2.o, 'hi');

    const s3 = newStruct('S2', {b: true, o: 'hi'});
    assert.isTrue(equals(s1, s3));
  });

  test('struct set', () => {
    const s1 = newStruct('S3', {b: true, o: 'hi'});
    const s2 = s1.setB(false);

    // TODO: assert throws on set wrong type
    assert.throws(() => {
      s1.setX(1);
    });

    const s3 = s2.setB(true);
    assert.isTrue(equals(s1, s3));

    const m = new StructMirror(s1);
    const s4 = m.set('b', false);
    assert.isTrue(equals(s2, s4));

    const s5 = s3.setO('bye');
    const s6 = new StructMirror(s3).set('o', 'bye');
    assert.isTrue(equals(s5, s6));

    // Changes the type
    const s7 = new StructMirror(s1).set('b', 42);
    assert.isTrue(equals(s7.type, makeStructType('S3', {
      b: numberType,
      o: stringType,
    })));

    // Adds a new field
    const s8 = new StructMirror(s1).set('x', 42);
    assert.isTrue(equals(s8.type, makeStructType('S3', {
      b: boolType,
      o: stringType,
      x: numberType,
    })));

    // Subtype
    const s9 = newStruct('', {l: new List([0, 1, false, true])});
    const s10 = new StructMirror(s9).set('l', new List([2, 3]));
    assert.isTrue(equals(s9.type, s10.type));
  });

  test('struct delete', () => {
    const s1 = newStruct('S', {b: true, o: 'hi'});
    const m1 = new StructMirror(s1);

    const s2 = m1.delete('notThere');
    assert.isTrue(equals(s1, s2));

    const s3 = m1.delete('o');
    const s4 = newStruct('S', {b: true});
    assert.isTrue(equals(s3, s4));

    const s5 = s1.deleteO();
    assert.isTrue(equals(s4, s5));

    const m5 = new StructMirror(s5);
    const s6 = m5.delete('b');
    const s7 = newStruct('S', {});
    assert.isTrue(equals(s6, s7));

    const s8 = s5.deleteB();
    assert.isTrue(equals(s7, s8));
  });

  test('createStructClass', () => {
    const typeA = makeStructType('A', {
      b: numberType,
      c: stringType,
    });
    const A = createStructClass(typeA);
    const a = new A({b: 1, c: 'hi'});
    assert.instanceOf(a, Struct);
    assert.instanceOf(a, A);
    assert.equal(a.b, 1);
    assert.equal(a.c, 'hi');
  });

  test('type validation', () => {
    const type = makeStructType('S1', {
      o: stringType,
      x: boolType,
    });

    assert.throws(() => {
      newStructWithType(type, ['hi', 1]);
    });
    assert.throws(() => {
      newStructWithType(type, [1]);
    });

    newStructWithType(type, ['hi', true]);
  });

  test('type validation cyclic', () => {
    // struct S {
    //   b: Bool
    //   l: List<S>
    // }
    const type = makeStructType('S', {
      b: boolType,
      l: makeListType(makeCycleType(0)),
    });

    const emptyList = new List([]);
    newStructWithType(type, [true, emptyList]);
    newStructWithType(type,
      [
        true,
        new List([newStructWithType(type, [false, emptyList])]),
      ]
    );

    assert.throws(() => {
      newStructWithType(type, [1]);
    });
    assert.throws(() => {
      newStructWithType(type, [true, 1]);
    });
  });

  function assertDiff(expect: string[], s1: Struct, s2: Struct) {
    const actual = structDiff(s1, s2);
    assert.deepEqual(expect, actual);
  }

  test('diff', async () => {
    const s1 = newStruct('', {a: true, b: 'hi', c: 4});

    assertDiff([], s1, newStruct('', {a: true, b: 'hi', c: 4}));
    assertDiff(['a', 'b'], s1, newStruct('', {a: false, b: 'bye', c: 4}));
    assertDiff(['b', 'c'], s1, newStruct('', {a: true, b: 'bye', c: 5}));
    assertDiff(['a', 'c'], s1, newStruct('', {a: false, b: 'hi', c: 10}));

    const s2 = newStruct('', {
      a: new List([0, 1]),
      b: new Map([['foo', false], ['bar', false]]),
      c: new Set([0, 1, 'foo']),
    });

    assertDiff([], s2, newStruct('', {
      a: new List([0, 1]),
      b: new Map([['foo', false], ['bar', false]]),
      c: new Set([0, 1, 'foo']),
    }));

    assertDiff(['a', 'b'], s2, newStruct('', {
      a: new List([1, 1]),
      b: new Map([['foo', false], ['bar', true]]),
      c: new Set([0, 1, 'foo']),
    }));

    assertDiff(['a', 'c'], s2, newStruct('', {
      a: new List([0]),
      b: new Map([['foo', false], ['bar', false]]),
      c: new Set([0, 1, 'bar']),
    }));

    assertDiff(['b', 'c'], s2, newStruct('', {
      a: new List([0, 1]),
      b: new Map([['boo', false], ['bar', true]]),
      c: new Set([0, 1, 'bar']),
    }));
  });

  test('escapeStructField', () => {
    const cases = [
      ['a', 'a'],
      ['AaZz19_', 'AaZz19_'],
      ['Q', 'Q51'],
      ['AQ1', 'AQ511'],
      ['INSPECTIONQ20STATUS', 'INSPECTIONQ5120STATUS'],
      ['$', 'Q24'],
      ['_content', 'Q5Fcontent'],
      ['Few ¢ents Short', 'FewQ20QC2A2entsQ20Short'],
      ['💩', 'QF09F92A9'],
    ];

    cases.forEach(c => {
      const [input, expected] = c;
      assert.equal(escapeStructField(input), expected);
    });
  });

  test('no override', () => {
    const s = newStruct('', {type: 'type', chunks: 'chunks', hash: 'hash'});
    assert.notEqual(s.type, 'type');
    assert.notEqual(s.chunks, 'chunks');
    assert.notEqual(s.hash, 'hash');

    assert.isTrue(equals(s.type,
        makeStructType('', {
          chunks: stringType,
          hash: stringType,
          type: stringType,
        })));
    assert.deepEqual(s.chunks, []);
    assert.instanceOf(s.hash, Hash);
  });

  test('no override - StructMirror', () => {
    const s = newStruct('', {type: 'type', chunks: 'chunks', hash: 'hash'});
    const m = new StructMirror(s);
    const names = [];
    const values = [];
    m.forEachField(f => {
      names.push(f.name);
      values.push(f.value);
      assert.isTrue(equals(f.type, stringType));
    });

    assert.deepEqual(names, ['chunks', 'hash', 'type']);
    assert.deepEqual(values, ['chunks', 'hash', 'type']);

    for (const n of names) {
      assert.equal(m.get(n), n);
    }

    for (const n of names) {
      const s2 = m.set(n, n);
      assert.isTrue(equals(s, s2));
    }
  });
});
